feat: initial code push

eneemann opened the pull request:
Here's the initial code we went over yesterday, plus loading secrets from a file.  It said I didn't have write permissions yet, so I created a fork.  I'll continue working on the items we talked about.

eneemann commented:
@steveoh I repointed this to the dev branch

steveoh commented on cool.py:
```diff
@@ -97,18 +97,12 @@ def download_tiles(col, row):
     img_4 = requests.get(f"{BASE_URL}/{col_num + 1}/{row_num + 1}").content
 
-    tile_images = [img_1, img_2, img_3, img_4]
+    tile_list = [img_1, img_2, img_3, img_4]
```
```suggestion
    return [img_1, img_2, img_3, img_4]
```

eneemann replied:
done

steveoh commented on cool.py:
```diff
@@ -97,18 +97,12 @@ def download_tiles(col, row):
+    tile_list = [img_1, img_2, img_3, img_4]
 
-    cv2_images = []
+    return tile_list
```
```suggestion
```

eneemann replied:
refactored this a bit

steveoh reviewed (commented):
**suggestion**: remove unnecessary variable

steveoh commented on cool.py:
```diff
@@ -0,0 +1,294 @@
+
+QUAD_WORD = None
+MODEL = None
+secrets = None
```
```suggestion
SECRETS = None
```

eneemann replied:
done

steveoh commented on cool.py:
```diff
@@ -0,0 +1,294 @@
+
+    if secrets is None:
+        logging.info('loading secrets')
+        secrets = SimpleNamespace(**_get_secrets())
```
**praise**: great use of simple namespace!

steveoh commented on cool.py:
```diff
@@ -0,0 +1,294 @@
+    Returns:
+        pytorch model: model ready for scanning
+    """
+    model_weight_path = Path(__file__).parent / 'tower_scout' / 'xl_250_best.pt'
+    yolov5_path = Path(__file__).parent / 'yolov5'
+    model = torch.hub.load(str(yolov5_path), 'custom', path=str(model_weight_path), source='local')
+    #: if statements (check that .pt file exists, folders, etc.)
```
**suggestion**: yup. get defensive here in case people didn't run the prerequisites.

steveoh commented on cool.py:
```diff
@@ -0,0 +1,294 @@
+    results = towerscout_model(image)
+    logging.info('time to scan image: %s',  format_time(perf_counter() - scan_start))
+
+    results.save()
```
**issue**: this should be removed or put behind a cli flag somehow

eneemann replied:
moved to CLI

steveoh commented on cool.py:
```diff
@@ -0,0 +1,294 @@
+
+    scan_start = perf_counter()
+    results = towerscout_model(image)
+    logging.info('time to scan image: %s',  format_time(perf_counter() - scan_start))
```
```suggestion
    logging.info('time to scan image: %s', format_time(perf_counter() - scan_start))
```

eneemann replied:
done

steveoh commented on cool.py:
```diff
@@ -0,0 +1,294 @@
+    zoom_level = 20
+
+    #: get upper-left coordinates of the primary tile (upper-left tile)
+    tile = mercantile.ul(int(col), int(row), zoom_level)
```
**question**: what happens if col or row are not integers?

eneemann replied:
it throws an error:
`TypeError: '<=' not supported between instances of 'int' and 'str'`

eneemann replied:
For this one, we're already casting the col/row values as ints in the function call.  Only strings of integers will be read into the row/col variables from our BQ table.  So what else should be done to be more defensive?

steveoh replied:
Generally speaking it's unsafe to assume that a database or user input will always be what you expect. This is the first time in this function the variables are cast to int. If it's not the the first time in the flow then the cast is unnecessary as you can ensure they are ints already. It's trivial to have a type check or a try except and exit early so you don't bomb out with an unhandled ValueError.

steveoh commented on cool.py:
```diff
@@ -0,0 +1,294 @@
+    results_df['x_centroid_3857'] = x + results_df['x_centroid_px'] * meters_per_pixel
+    results_df['y_centroid_3857'] = y - results_df['y_centroid_px'] * meters_per_pixel
+
+    print(results_df.head().to_string())
```
**chore**: remove debugging

eneemann replied:
done

steveoh reviewed (commented):
please add the following items from the agrc/python repo

- `.github` folder
- `.vscode` folder
- `.coveragerc` file
- `.editorconfig` file
- `LICENSE` file
- `pylintrc` file
- `setup.cfg` file

eneemann commented on cool_cli.py:
```diff
@@ -0,0 +1,76 @@
+Examples:
+    python cool_cli.py download tiles 198259 394029 --save-to=./data --mosaic
+    python cool_cli.py download tiles 198402 394067 --mosaic --detect
+    python cool_cli.py download tiles 198402 394067 --mosaic --detect --parse-results
```
I'll let you review the latest changes

steveoh replied:
Yeah I like what you did.

eneemann commented:
Okay @steveoh, tear it apart. I'm kinda stuck on how to organize the CLI.  It doesn't make sense to me to run the mosaic command alone, because we'd have to build a list of images and read them in or refactor the function, so I packaged that as an option on download. I changed parse-results to locate-results because that function is creating the spatial info (but we need to specify the col/row to enable a mosaic image to be located).  The process-tile command runs through the whole chain to this point.

I made some of the changes you requested so far. It took a while, but I think I've refactored the tile downloads into something more robust with timeouts and retries, borrowing a bunch from Jake's Palletjack code.

For the BGR/RGB issue, it was only affecting in-memory np.ndarrays, so I created a function to reorder the color channels when a np.ndarray is detected. I think the cv2 function knows it has to write out as RGB when saving JPGs, so pushing a Path object for the JPG through the PyTorch model wasn't impacted by the reversed colors.

Still more to do (including tests), but making a little progress.

steveoh commented on cool.py:
```diff
@@ -65,41 +64,165 @@ def convert_to_cv2_image(image):
+        return None
+
+    #: if we haven't returned, raise an error to trigger _retry
+    raise RuntimeError(f"Did not receive a valid tile download response; status code: {response.status_code}")
```
https://requests.readthedocs.io/en/latest/api/#requests.Response.raise_for_status

steveoh replied:
I prefer scott's masquerade code for [retries](https://github.com/agrc/masquerade/blob/main/src/masquerade/providers/web_api.py#L67) as it uses built it library code. His `make_request` method could use some exception handling e.g. 

```py
import requests

try:
    response = requests.get(url, timeout=5)
    response.raise_for_status() # Raise an exception if the status code is not 200 OK
except requests.exceptions.Timeout:
    # Timeout occurred
    print("Timeout error occurred")
except requests.exceptions.RequestException as e:
    # Other error occurred
    print("Error occurred: {}".format(e))
```

But if you were to combine the two I think you'd have a good reference implementation that masquerade and palletjack could benefit from.

steveoh commented on cool.py:
```diff
@@ -141,23 +131,31 @@ def get_tile(url):
+
+    try:
+        response = session.get(url, timeout=5)
+        response.raise_for_status()  #: raise an exception if the status code is not 200 OK
```
**thought**: you might need to be careful here as a 404 will throw but we probably just want to skip it from the mosaic unless all of the tiles 404 and you just have a white image.

steveoh commented on cool.py:
```diff
@@ -141,23 +131,31 @@ def get_tile(url):
+        response.raise_for_status()  #: raise an exception if the status code is not 200 OK
+    except requests.exceptions.Timeout:
+        #: timeout occurred
+        logging.debug("Timeout error occurred")
```
**chore**: elevate this to an error or at least a warning.

steveoh commented on cool.py:
```diff
@@ -141,23 +131,31 @@ def get_tile(url):
+
+    except requests.exceptions.RequestException as e:
+        #: other error occurred
+        logging.debug("Error occurred: %s", e)
```
**chore**: elevate this to a warning or higher
